src/node_dns.cc

```diff
@@ -0,0 +1,471 @@
+                              server->Get(env->context(), 0).ToLocalChecked());
+    if (!ConvertAddress(env->isolate(), *address, dict)) {
+      return;
+    }
+    uint32_t port = server->Get(env->context(), 1)
+                        .ToLocalChecked()
+                        ->Uint32Value(env->context())
+                        .ToChecked();
+    Local<Value> tls_hostname = server->Get(env->context(), 2).ToLocalChecked();
+    if (!tls_hostname->IsUndefined()) {
+      GETDNS(dict_set_int, env->isolate(), dict, "tls_port", port);
+      String::Utf8Value hostname(env->isolate(), tls_hostname);
+      GETDNS(dict_util_set_string,
+             env->isolate(),
+             dict,
+             "tls_auth_name",
+             *hostname);
+    } else {
+      GETDNS(dict_set_int, env->isolate(), dict, "port", port);
+    }
+    GETDNS(list_set_dict, env->isolate(), list, i, dict);
+  }
+
+  GETDNS(context_set_upstream_recursive_servers,
+         env->isolate(),
+         dns->context_,
+         list);
+}
+
+// static
+void DNSWrap::GetUpstreamRecursiveServers(
+    const FunctionCallbackInfo<Value>& args) {
+  DNSWrap* dns;
+  ASSIGN_OR_RETURN_UNWRAP(&dns, args.This());
+  Environment* env = dns->env();
+
+  getdns_list* list;
+  auto cleanup = OnScopeLeave([&]() {
+    getdns_list_destroy(list);
+  });
+  GETDNS(context_get_upstream_recursive_servers,
+         env->isolate(),
+         dns->context_,
+         &list);
+
+  char* json = getdns_print_json_list(list, 0);
+  Local<String> result =
+      String::NewFromUtf8(env->isolate(), json).ToLocalChecked();
+  free(json);
+
+  args.GetReturnValue().Set(result);
+}
+
+// static
+void DNSWrap::SetDNSTransportList(const FunctionCallbackInfo<Value>& args) {
+  DNSWrap* dns;
+  ASSIGN_OR_RETURN_UNWRAP(&dns, args.This());
+  Environment* env = dns->env();
+
+  CHECK(args[0]->IsArray());
+  Local<Array> transports = args[0].As<Array>();
+  uint32_t length = transports->Length();
+
+  std::vector<getdns_transport_list_t> transport_list(length);
+  for (uint32_t i = 0; i < length; i += 1) {
+    getdns_transport_list_t transport =
+        static_cast<getdns_transport_list_t>(transports->Get(env->context(), i)
+                                                 .ToLocalChecked()
+                                                 ->Uint32Value(env->context())
+                                                 .ToChecked());
+    transport_list[i] = transport;
+  }
+
+  GETDNS(context_set_dns_transport_list,
+         env->isolate(),
+         dns->context_,
+         length,
+         &transport_list[0]);
+}
+
+// static
+void DNSWrap::CancelAllTransactions(const FunctionCallbackInfo<Value>& args) {
+  DNSWrap* dns;
+  ASSIGN_OR_RETURN_UNWRAP(&dns, args.This());
+  std::vector<getdns_transaction_t> transactions;
+  for (auto& t : dns->transactions_) {
+    transactions.push_back(t.first);
+  }
+  for (auto& t : transactions) {
+    GETDNS(cancel_callback, dns->env()->isolate(), dns->context_, t);
+  }
```

Comment:
Can the cancel be processed in the first for loop to avoid the double iteration?

Reply:
Cancelling them runs DNSWrap::RejectTransaction which mutates `transactions_` so a copy is needed.